{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"#1 - Understanding Poly.ipynb","provenance":[{"file_id":"1GWFFYUcduoykRm08R6Jq7OacGQmDF1BS","timestamp":1634754196930},{"file_id":"1Ts_naBl0P_KfVvwC90Fb-JGWDeP6qjrV","timestamp":1634234790466},{"file_id":"14Bh2CljYULhD03AKCNGlJ3YgM9T5mcRa","timestamp":1627231003334},{"file_id":"1fZTAiBlR5eicaMu4nhhtfeJZEGMUAMHD","timestamp":1626732496856},{"file_id":"1NGDMQi-OGV5-XAbOgz87JWtSc4l2hHyh","timestamp":1626622594804}],"collapsed_sections":["3STJpf80Bl9Z"]},"kernelspec":{"display_name":"Go","name":"gophernotes"}},"cells":[{"cell_type":"markdown","metadata":{"id":"dW2UGT4s32-x"},"source":["# ***Poly*, a go package for organism engineering** \n","\n","***Poly*** is a fast and modern project created by [**@TimothyStiles**](https://twitter.com/TimothyStiles) to resolve different problems found while engineering living things. From designing primers for barcoding to CDS fixing, Poly is trying to be the most complete, open and well-used computational synthetic biology tool.\n","\n","\n","---\n","\n","\n","Poly is our key tool for the software. It was a planned decision to build workflows that integrate with this package, to show ways to use the package, as well as create some new features; therefore, it is very important that you understand how the Poly package works and what its structure is in general before you begin manipulating it. Thus, we created this brief overview of Poly, its sub-packages, and a collection of use cases. We strongly recommend that you do the tutorials in the order they appear.\n"]},{"cell_type":"markdown","metadata":{"id":"PZJDiu_W3_FT"},"source":["# Configurations for this tutorial\n","\n"]},{"cell_type":"markdown","metadata":{"id":"Xqjf_sf0Am-X"},"source":["\n","First let's run some important settings so you can run this tutorial successfully. \n"]},{"cell_type":"markdown","metadata":{"id":"96PZqanZBAle"},"source":["Colab notebooks use python kernels to run each cell. However, because ***Poly*** is written in **Go language (golang)**, we need to install and configure some things in colab to make feasible run something in go lang here.\n","\n","### **1. In order to start the golang environment, run the line below:**"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"caT1iMfshw62","executionInfo":{"status":"ok","timestamp":1634769301186,"user_tz":180,"elapsed":134782,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"8f3f3499-4c0e-47fc-f4c1-5ce5ee42fa54"},"source":["# this process may take a few minutes\n","!add-apt-repository ppa:longsleep/golang-backports -y\n","!apt update\n","!apt install golang-go\n","%env GOPATH=/root/go\n","!go get -u github.com/gopherdata/gophernotes\n","!cp ~/go/bin/gophernotes /usr/bin/\n","!npx degit gopherdata/gophernotes/kernel \\\n","     /usr/local/share/jupyter/kernels/gophernotes"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["\r0% [Working]\r            \rGet:1 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]\n","\r0% [Connecting to archive.ubuntu.com (91.189.88.152)] [1 InRelease 14.2 kB/88.7\r                                                                               \rGet:2 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]\n","Ign:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n","Get:4 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]\n","Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n","Get:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release [696 B]\n","Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n","Get:8 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release.gpg [836 B]\n","Hit:9 http://archive.ubuntu.com/ubuntu bionic InRelease\n","Get:10 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]\n","Get:11 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ Packages [69.5 kB]\n","Hit:12 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n","Hit:13 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease\n","Get:14 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]\n","Get:15 http://security.ubuntu.com/ubuntu bionic-security/restricted amd64 Packages [626 kB]\n","Hit:16 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n","Get:17 http://security.ubuntu.com/ubuntu bionic-security/universe amd64 Packages [1,434 kB]\n","Get:18 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [2,385 kB]\n","Get:19 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic InRelease [15.4 kB]\n","Get:21 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Packages [753 kB]\n","Get:22 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main Sources [1,809 kB]\n","Get:23 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [2,213 kB]\n","Get:24 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main amd64 Packages [926 kB]\n","Get:25 http://archive.ubuntu.com/ubuntu bionic-updates/restricted amd64 Packages [664 kB]\n","Get:26 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [2,833 kB]\n","Get:27 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 Packages [6,639 B]\n","Fetched 14.0 MB in 4s (3,285 kB/s)\n","Reading package lists... Done\n","Hit:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease\n","Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n","Hit:3 http://security.ubuntu.com/ubuntu bionic-security InRelease\n","Hit:4 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease\n","Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n","Hit:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release\n","Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n","Hit:8 http://archive.ubuntu.com/ubuntu bionic InRelease\n","Hit:9 http://archive.ubuntu.com/ubuntu bionic-updates InRelease\n","Hit:10 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n","Hit:11 http://archive.ubuntu.com/ubuntu bionic-backports InRelease\n","Hit:12 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease\n","Hit:13 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n","Hit:14 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic InRelease\n","Reading package lists... Done\n","Building dependency tree       \n","Reading state information... Done\n","67 packages can be upgraded. Run 'apt list --upgradable' to see them.\n","Reading package lists... Done\n","Building dependency tree       \n","Reading state information... Done\n","The following additional packages will be installed:\n","  golang-1.17-go golang-1.17-src golang-src\n","Suggested packages:\n","  bzr | brz mercurial subversion\n","The following NEW packages will be installed:\n","  golang-1.17-go golang-1.17-src golang-go golang-src\n","0 upgraded, 4 newly installed, 0 to remove and 67 not upgraded.\n","Need to get 71.8 MB of archives.\n","After this operation, 423 MB of additional disk space will be used.\n","Get:1 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-1.17-src all 1.17.2-1longsleep1+focal [15.2 MB]\n","Get:2 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-1.17-go amd64 1.17.2-1longsleep1+focal [56.6 MB]\n","Get:3 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-src amd64 2:1.17~1longsleep1 [5,056 B]\n","Get:4 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-go amd64 2:1.17~1longsleep1 [24.1 kB]\n","Fetched 71.8 MB in 5s (15.4 MB/s)\n","Selecting previously unselected package golang-1.17-src.\n","(Reading database ... 155047 files and directories currently installed.)\n","Preparing to unpack .../golang-1.17-src_1.17.2-1longsleep1+focal_all.deb ...\n","Unpacking golang-1.17-src (1.17.2-1longsleep1+focal) ...\n","Selecting previously unselected package golang-1.17-go.\n","Preparing to unpack .../golang-1.17-go_1.17.2-1longsleep1+focal_amd64.deb ...\n","Unpacking golang-1.17-go (1.17.2-1longsleep1+focal) ...\n","Selecting previously unselected package golang-src:amd64.\n","Preparing to unpack .../golang-src_2%3a1.17~1longsleep1_amd64.deb ...\n","Unpacking golang-src:amd64 (2:1.17~1longsleep1) ...\n","Selecting previously unselected package golang-go.\n","Preparing to unpack .../golang-go_2%3a1.17~1longsleep1_amd64.deb ...\n","Unpacking golang-go (2:1.17~1longsleep1) ...\n","Setting up golang-1.17-src (1.17.2-1longsleep1+focal) ...\n","Setting up golang-src:amd64 (2:1.17~1longsleep1) ...\n","Setting up golang-1.17-go (1.17.2-1longsleep1+focal) ...\n","Setting up golang-go (2:1.17~1longsleep1) ...\n","Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n","env: GOPATH=/root/go\n","go: downloading github.com/gopherdata/gophernotes v0.7.3\n","go: downloading github.com/cosmos72/gomacro v0.0.0-20210624153544-b4935e406a41\n","go: downloading github.com/go-zeromq/zmq4 v0.13.0\n","go: downloading github.com/gofrs/uuid v4.0.0+incompatible\n","go: downloading github.com/gofrs/uuid v4.1.0+incompatible\n","go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1\n","go: downloading github.com/peterh/liner v1.2.1\n","go: downloading github.com/go-zeromq/goczmq/v4 v4.2.2\n","go: downloading github.com/go-zeromq/goczmq v4.1.0+incompatible\n","go: downloading golang.org/x/sync v0.0.0-20201020160332-67f06af15bc9\n","go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c\n","go: downloading golang.org/x/tools v0.1.0\n","go: downloading golang.org/x/tools v0.1.7\n","go: downloading github.com/mattn/go-runewidth v0.0.12\n","go: downloading github.com/mattn/go-runewidth v0.0.13\n","go: downloading golang.org/x/sys v0.0.0-20210119212857-b64e53b001e4\n","go: downloading golang.org/x/sys v0.0.0-20211020174200-9d6173849985\n","go: downloading github.com/rivo/uniseg v0.1.0\n","go: downloading github.com/rivo/uniseg v0.2.0\n","go: downloading golang.org/x/mod v0.3.0\n","go: downloading golang.org/x/mod v0.5.1\n","go get: installing executables with 'go get' in module mode is deprecated.\n","\tUse 'go install pkg@version' instead.\n","\tFor more information, see https://golang.org/doc/go-get-install-deprecation\n","\tor run 'go help get' or 'go help install'.\n","\u001b[K\u001b[?25hnpx: installed 1 in 3.563s\n","\u001b[36m> cloned \u001b[1mgopherdata/gophernotes\u001b[22m#\u001b[1mHEAD\u001b[22m to /usr/local/share/jupyter/kernels/gophernotes\u001b[39m\n"]}]},{"cell_type":"markdown","metadata":{"id":"cZwRFE9UAGIs"},"source":["### **2. Download important data to run this tutorial**"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"wHmBJ418_5tX","executionInfo":{"status":"ok","timestamp":1634769320860,"user_tz":180,"elapsed":19726,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"c4f7a550-4f6f-4503-b21c-33b9c338fcbc"},"source":["!rm -rf $GOPATH/pkg/mod/github.com/!open-!science-!global\n","!rm -rf $GOPATH/pkg/mod/cache/download/github.com/!open-!science-!global\n","!go get -u github.com/Open-Science-Global/poly@e3e1c61"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["go: downloading github.com/Open-Science-Global/poly v0.11.3\n"]}]},{"cell_type":"code","metadata":{"id":"KPWnoUlq7TSr"},"source":["!wget https://raw.githubusercontent.com/Open-Science-Global/friendzymes-cookbook/main/data/understanding-poly/base.fasta\n","!wget https://raw.githubusercontent.com/Open-Science-Global/friendzymes-cookbook/main/data/understanding-poly/puc19.gbk"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"3STJpf80Bl9Z"},"source":["### **3. Connect Colab Notebook to your GDrive (not required)**"]},{"cell_type":"markdown","metadata":{"id":"ftGzhtH5AOZM"},"source":["The previous code will download the files temporarily. If you want to download them to a folder on your drive and save it for later analysis or if you are already using this notebook to run your own files, you should connect your Google Drive to Colab and you will be able to access, read and save files permanently. \n","So, if you prefer, you can do this with the code line below:\n","\n"]},{"cell_type":"code","metadata":{"id":"sh661uGn_8R1"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"GbPLWTJ05uH4"},"source":["## After running these steps, click on **Runtime** in the menu bar & **Change Runtime Type** to Go, if it hasn't changed yet.\n","\n","This will make Colab starting use a open source go kernel called Gopher Notes.\n","\n","Now we are ready to work."]},{"cell_type":"markdown","metadata":{"id":"euQ4kI-79Rbk"},"source":["# **Importing Poly**"]},{"cell_type":"markdown","metadata":{"id":"J7H6s_dmAh1Z"},"source":["The first thing to know is that we created a fork of Poly for our Friendzymes repository to implements new features and improvements fast. You could take a look in the project [**here**](https://github.com/Open-Science-Global/poly).\n"]},{"cell_type":"markdown","metadata":{"id":"INRbBk-DAsJn"},"source":["## **A tour of Poly's key subpackages**"]},{"cell_type":"markdown","metadata":{"id":"fjLWExeRDdrR"},"source":["Poly has features divided by subpackages. Each subpackage has functions and data structures related to some topic, e.g. clone, synthesis, primers, random, transform, io and others.\n","\n","To import a subpackage you could use like the example below for the **subpackage random** and golang will know what you want."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"jfA0N5x0_mCb","executionInfo":{"status":"ok","timestamp":1634754562501,"user_tz":180,"elapsed":9672,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"54866fac-4c87-4145-c0e2-51d3fca6820a"},"source":["import \"github.com/Open-Science-Global/poly/random\"\n","\n","// example of use of the random subpackage\n","randomSequence, err := random.ProteinSequence(100, 1)\n","randomSequence"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["MCIIYCWGATARNDLJRNGVHSHJWIIIJMSCJINLTVNGHQMREQIWFYQVCLYAGJWESNMGTHJCDEHETDWIRVETAEQVQCYQENDWTHIAEPE*"]},"metadata":{},"execution_count":1}]},{"cell_type":"markdown","metadata":{"id":"CQ0PTqceAx-x"},"source":["## IO (Input Output)"]},{"cell_type":"markdown","metadata":{"id":"VdnWr10SA7eP"},"source":["IO is the subpackage related to reading and writing of files extensions commonly used in synthetic biology. Poly could deal with the following file extensions:\n","\n","\n","\n","*   **Genbank**\n","*   **FASTA**\n","*   **Poly Json** (a new proposal to parse genetic sequences)\n","*   **GFF**\n","*   **Rebase**\n","*   **Uniprot**\n","\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"IEGZRpwVGJlS","executionInfo":{"status":"ok","timestamp":1634754587356,"user_tz":180,"elapsed":17243,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"36076159-3bc4-417a-83d5-4f3847b9e0b6"},"source":["import (\n","  \"fmt\"\n","  \"github.com/Open-Science-Global/poly/io/fasta\"\n","  \"github.com/Open-Science-Global/poly/io/genbank\"\n",")\n","\n","fmt.Println(\"#Read a fasta file and show the sequences:\")\n","\n","bases := fasta.Read(\"./base.fasta\")\n","for i, v := range(bases) {\n","  fmt.Println(v)\n","}\n","\n","fmt.Println()\n","fmt.Println(\"#Read a genbank file and show the sequences:\")\n","\n","puc19 := genbank.Read(\"./puc19.gbk\")\n","fmt.Println(\"GenBank Sequence\", \"|\", puc19.Sequence)\n","\n","for i, feature := range(puc19.Features) {\n","  if (feature.Type == \"CDS\") {\n","    fmt.Println(\"CDS\", \"|\", feature.Attributes[\"product\"], \"|\", feature.Attributes[\"translation\"])\n","  }\n","}"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["#Read a fasta file and showing the sequences:\n","{ }\n","\n","#Read a genbank file and showing the sequences:\n","GenBank Sequence | \n"]}]},{"cell_type":"markdown","metadata":{"id":"8ECT8d12A-k4"},"source":["## Transform"]},{"cell_type":"markdown","metadata":{"id":"giorDEcSj903"},"source":["\n","The transform subpackage deal with some *transformation* as get a **complement** or **reverse complement** of some DNA sequence. It also have another subpackage called ***codons*** to deal with **translation** and **codon optimization**."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"IlGjS08xU5KL","executionInfo":{"status":"ok","timestamp":1634754721200,"user_tz":180,"elapsed":11,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"625ded69-6fa1-445a-a35a-c683c2eee23f"},"source":["import (\n","  \"fmt\"\n","  \"github.com/Open-Science-Global/poly/transform\"\n","  \"github.com/Open-Science-Global/poly/transform/codon\"\n",")\n","\n","seq := \"GATTACA\"\n","fmt.Println(\"Sequence:\", seq, \"Complement:\", transform.Complement(seq), \"Reverse Complement:\", transform.ReverseComplement(seq))\n","\n","gfp := \"ATGGCTAGCAAAGGAGAAGAACTTTTCACTGGAGTTGTCCCAATTCTTGTTGAATTAGATGGTGATGTTAATGGGCACAAATTTTCTGTCAGTGGAGAGGGTGAAGGTGATGCTACATACGGAAAGCTTACCCTTAAATTTATTTGCACTACTGGAAAACTACCTGTTCCATGGCCAACACTTGTCACTACTTTCTCTTATGGTGTTCAATGCTTTTCCCGTTATCCGGATCATATGAAACGGCATGACTTTTTCAAGAGTGCCATGCCCGAAGGTTATGTACAGGAACGCACTATATCTTTCAAAGATGACGGGAACTACAAGACGCGTGCTGAAGTCAAGTTTGAAGGTGATACCCTTGTTAATCGTATCGAGTTAAAAGGTATTGATTTTAAAGAAGATGGAAACATTCTCGGACACAAACTCGAGTACAACTATAACTCACACAATGTATACATCACGGCAGACAAACAAAAGAATGGAATCAAAGCTAACTTCAAAATTCGCCACAACATTGAAGATGGATCCGTTCAACTAGCAGACCATTATCAACAAAATACTCCAATTGGCGATGGCCCTGTCCTTTTACCAGACAACCATTACCTGTCGACACAATCTGCCCTTTCGAAAGATCCCAACGAAAAGCGTGACCACATGGTCCTTCTTGAGTTTGTAACTGCTGCTGGGATTACACATGGCATGGATGAGCTCTACAAATAA\"\n","fmt.Println(\"You could translate a sequence using codon.Translate passing the sequence AND a referenced codon table\")\n","translation, _ := codon.Translate(gfp, codon.GetCodonTable(11))\n","fmt.Println(\"GFP Translated:\", translation)\n","\n","fmt.Println()\n","fmt.Println(\"You could optimize a sequence using codon.Optimize passing the sequence AND optimized codon table\")\n","optimized, _ := codon.Optimize(gfp, codon.GetCodonTable(11).OptimizeTable(gfp))\n","fmt.Println(\"GFP Codon Optimization:\", optimized)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Sequence: GATTACA Complement: CTAATGT Reverse Complement: TGTAATC\n","You could translate a sequence using codon.Translate passing the sequence AND a referenced codon table\n","GFP Translated: MASKGEELFTGVVPILVELDGDVNGHKFSVSGEGEGDATYGKLTLKFICTTGKLPVPWPTLVTTFSYGVQCFSRYPDHMKRHDFFKSAMPEGYVQERTISFKDDGNYKTRAEVKFEGDTLVNRIELKGIDFKEDGNILGHKLEYNYNSHNVYITADKQKNGIKANFKIRHNIEDGSVQLADHYQQNTPIGDGPVLLPDNHYLSTQSALSKDPNEKRDHMVLLEFVTAAGITHGMDELYK*\n","\n","You could optimize a sequence using codon.Optimize passing the sequence AND optimized codon table\n","GFP Codon Optimization: GCTACAGGAGGCTGCACTGCAGGGTGCGCTGCCGCAGGAGGTGCTGGAGCTGCTGGAGCAGCCTGCACTACTACAACATGCGCTTGCACAGGAGGAGCTGGCACTACTGGAACATGCTGCTGCGCTGCAACAACCTGCACTACCGGAACAACGGGAGCAGCTACTACTGCCGGTGCTACTGGAGGTACCGGAGCTACTGGAACAACTGCTGCTACAGGAGGAGGGTGCGCATGCGCTGCTGCAACGACTACTACTTGCACCGGAACATGCGCAGGTACTGGAGGGGCAGGTGCCGGTGGCGGTACAGGAGCTGCAGGAGGAACTGGCGCTACGGGTTGCACTGCTTGCGCTACCGCTTGCGGCGGTGCTGCTGCTGGTTGCACTACAGCTTGCTGCTGCACGACTGCTGCTGCTACAACGACAGCCACTACTACTGGCTGCGCTTGCACAGCTTGCACGGGAGGAGCCGCAGCCGCTTGCACTGCTTGCTGCACAGGTACTACTTGCTGCGCCACTGGAGGATGCTGCGCTGCTTGCGCTTGCACAACTGGGACCTGCGCTTGCACGGCTTGCACAACTACTTGCACCTGCACCACTGCTACAGGTGGCACTGGAACTACATGCGCCGCTACAGGGTGCACTACAACTACGTGCTGCTGCGGAACAACTGCTACCTGCTGCGGAGGAGCTACTTGCGCCACTGCAACTGGTGCTGCTGCTTGCGGCGGATGCGCAACTGGGGCATGCACTACAACGACTACGTGCGCTGCTGGTGCTGGAACTGGTTGCTGCGCCACTGGGTGCTGCTGCGGAGCAGCTGGAGGTACGACCGCCACGGGTACAGCCTGCGCTGGTGGTGCTGCTTGCGGGTGCGCTTGCACTGCTACCGCCACCTGCACGACTACATGCGCTGCCGCCGGAGCAACAGGAGCTTGCGGCGGTGGAGCAGCCTGCACGGCTTGCGCTGCTGGGGCTTGCGGGTGCGGAACTGGTTGCACTGGTGCTGCCGGTACTTGCGCTGCTGGAACTACAACTGGAGCTGCTGGTGGGACAGGCGCTACAGCTTGCTGCTGCACGACTGGGACCACTGCTGCAACTTGCGGCACGGCTACCTGCGGAGCTGGAACTACTGCTGCAGCTGCAGGAGGTACTGCTACAACAGGAGCCACAACGACTACAGCAGCCGCTGGGGCTGCAGGAGCTACTGGCGGTGCTGCTGCCTGCGCTACAACTTGCACATGCGGCGGAGCTTGCGCTTGCGCTGCAGCTTGCACATGCGGAGCTGGAACTGCATGCGCTGCTTGCACAGCCACGGCTGCTTGCACTTGCGCTTGCGCATGCGCTGCCACCGGTACAGCTACCGCTTGCGCAACCTGCGCTTGCGGCGGCTGCGCCGGAGCCTGCGCTGCTGCTTGCGCCGCTGCTGCCGGTGCTGCTACTGGTGGCGCAGCAACTTGCGCTGCTGCCGGTTGCACTGCTGCTTGCACAACTTGCGCAGCCGCTGCAACTACATGCGGCTGCTGCGCTTGCGCTGCTTGCGCTACAACCGGCGCCGCTGGTGCTACAGGGGGTGCTACATGCTGCGGCACTACATGCGCTGCTTGCACAGCTGGTTGCGCTGGAGCTTGCTGCGCCACTACTGCTACTTGCGCCGCATGCGCCGCTGCTGCAACAGCTTGCACGTGCTGCGCTGCTACTACGGGAGGATGCGGAGCTACTGGTGGATGCTGCTGCACGGGGACTTGCTGCACGACCACTACAGCATGCTGCGCTGGAGCCTGCGCTGCCTGCTGCGCTACTACAGCCTGCTGCACTGGTACTTGCGGGGCTTGCGCTTGCGCCGCTACTTGCACTGGATGCTGCTGCACAACAACTTGCGGTGCTGCTGCTGGTGCTACTTGCTGCTGCGCAGCTTGCGGGGCCGCAGCAGCTGGATGCGGAACTGGAGCTTGCTGCGCTTGCGCTACTGGAGGAACTTGCTGCACGACGTGCACGACGGGAGCTGGTACAACTACAGGTACAGCTGCATGCACCGGATGCACCGGATGCACAGGAGGTGGTGCAACTACTGCTTGCGCCTGCGCTACTGGTGGATGCGCCACCGGAGGAGCTACTGGGGCTGGATGCACTTGCACAGCTTGCGCTGCTGCTACCGCTGCC\n"]},{"output_type":"execute_result","data":{"text/plain":["2185 <nil>"]},"metadata":{},"execution_count":4}]},{"cell_type":"markdown","metadata":{"id":"4jvlGGghZBzx"},"source":["## Clone"]},{"cell_type":"markdown","metadata":{"id":"n33r5hC0ZIpz"},"source":["The Clone subpackage has two main functionalies: \n","- 1) a function that simulates restriction digestion and returns a list of fragments (linear and circular sequences)\n","  \n","- **Important:** every time we have \"false\" as parameter within the variable it means that this is not from a circular sequence. For example: `cloned := clone.Part{mySeq1 + bsaiBindingSite + mySeq2, false} `. When we have \"true\" it means that it comes from a circular sequence like a plasmid."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"WGTT4MGYdByJ","executionInfo":{"status":"ok","timestamp":1634755467553,"user_tz":180,"elapsed":397,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"ff2b4d0e-d750-4afb-aa7e-b1b7ba96a81b"},"source":["import (\n","  \"fmt\"\n","  \"github.com/Open-Science-Global/poly/clone\"\n",")\n","bsaiBindingSite := \"GGTCTCAATGC\"\n","mySeq1 := \"ATATATATATATATAT\"\n","mySeq2 := \"GCGCGCGCGCGCGCGCGCGC\"\n","cloned := clone.Part{mySeq1 + bsaiBindingSite + mySeq2, false} // false means it's not from a circular sequence\n","cloned"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["{ATATATATATATATATGGTCTCAATGCGCGCGCGCGCGCGCGCGCGC false}"]},"metadata":{},"execution_count":9}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kkF_hjEUAkib","executionInfo":{"status":"ok","timestamp":1634755468883,"user_tz":180,"elapsed":6,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"02096406-8911-426e-c883-44c8a964d2de"},"source":["fragments, err := clone.CutWithEnzymeByName(cloned, false, \"BsaI\")\n","\n","fragments"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["[{GCGCGCGCGCGCGCGCGCGC ATGC } {ATATATATATATATATGGTCTCA  ATGC}]"]},"metadata":{},"execution_count":10}]},{"cell_type":"markdown","metadata":{"id":"0LoIqIuWBPkS"},"source":["- 2) a Golden Gate Assembly simulator\n","\n","The variable `popen` is the sequence of the plasmid (note that it is with the \"true\" parameter), `fragment1` and `fragment2` are the parts I want to clone. Then this is passed to the `GoldenGate()` function with the indication of the enzyme we want to simulate, which in this case is BbsI.\n"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"UXz-LGGNgF3Q","executionInfo":{"status":"ok","timestamp":1634755604159,"user_tz":180,"elapsed":347,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"bbf97820-d63b-46e4-8bc9-101e0d363438"},"source":["var popen = clone.Part{\"TAACTATCGTCTTGAGTCCAACCCGGTAAGACACGACTTATCGCCACTGGCAGCAGCCACTGGTAACAGGATTAGCAGAGCGAGGTATGTAGGCGGTGCTACAGAGTTCTTGAAGTGGTGGCCTAACTACGGCTACACTAGAAGAACAGTATTTGGTATCTGCGCTCTGCTGAAGCCAGTTACCTTCGGAAAAAGAGTTGGTAGCTCTTGATCCGGCAAACAAACCACCGCTGGTAGCGGTGGTTTTTTTGTTTGCAAGCAGCAGATTACGCGCAGAAAAAAAGGATCTCAAGAAGGCCTACTATTAGCAACAACGATCCTTTGATCTTTTCTACGGGGTCTGACGCTCAGTGGAACGAAAACTCACGTTAAGGGATTTTGGTCATGAGATTATCAAAAAGGATCTTCACCTAGATCCTTTTAAATTAAAAATGAAGTTTTAAATCAATCTAAAGTATATATGAGTAAACTTGGTCTGACAGTTACCAATGCTTAATCAGTGAGGCACCTATCTCAGCGATCTGTCTATTTCGTTCATCCATAGTTGCCTGACTCCCCGTCGTGTAGATAACTACGATACGGGAGGGCTTACCATCTGGCCCCAGTGCTGCAATGATACCGCGAGAACCACGCTCACCGGCTCCAGATTTATCAGCAATAAACCAGCCAGCCGGAAGGGCCGAGCGCAGAAGTGGTCCTGCAACTTTATCCGCCTCCATCCAGTCTATTAATTGTTGCCGGGAAGCTAGAGTAAGTAGTTCGCCAGTTAATAGTTTGCGCAACGTTGTTGCCATTGCTACAGGCATCGTGGTGTCACGCTCGTCGTTTGGTATGGCTTCATTCAGCTCCGGTTCCCAACGATCAAGGCGAGTTACATGATCCCCCATGTTGTGCAAAAAAGCGGTTAGCTCCTTCGGTCCTCCGATCGTTGTCAGAAGTAAGTTGGCCGCAGTGTTATCACTCATGGTTATGGCAGCACTGCATAATTCTCTTACTGTCATGCCATCCGTAAGATGCTTTTCTGTGACTGGTGAGTACTCAACCAAGTCATTCTGAGAATAGTGTATGCGGCGACCGAGTTGCTCTTGCCCGGCGTCAATACGGGATAATACCGCGCCACATAGCAGAACTTTAAAAGTGCTCATCATTGGAAAACGTTCTTCGGGGCGAAAACTCTCAAGGATCTTACCGCTGTTGAGATCCAGTTCGATGTAACCCACTCGTGCACCCAACTGATCTTCAGCATCTTTTACTTTCACCAGCGTTTCTGGGTGAGCAAAAACAGGAAGGCAAAATGCCGCAAAAAAGGGAATAAGGGCGACACGGAAATGTTGAATACTCATACTCTTCCTTTTTCAATATTATTGAAGCATTTATCAGGGTTATTGTCTCATGAGCGGATACATATTTGAATGTATTTAGAAAAATAAACAAATAGGGGTTCCGCGCACCTGCACCAGTCAGTAAAACGACGGCCAGTAGTCAAAAGCCTCCGACCGGAGGCTTTTGACTTGGTTCAGGTGGAGTGGGAGTAgtcttcGCcatcgCtACTAAAagccagataacagtatgcgtatttgcgcgctgatttttgcggtataagaatatatactgatatgtatacccgaagtatgtcaaaaagaggtatgctatgaagcagcgtattacagtgacagttgacagcgacagctatcagttgctcaaggcatatatgatgtcaatatctccggtctggtaagcacaaccatgcagaatgaagcccgtcgtctgcgtgccgaacgctggaaagcggaaaatcaggaagggatggctgaggtcgcccggtttattgaaatgaacggctcttttgctgacgagaacagggGCTGGTGAAATGCAGTTTAAGGTTTACACCTATAAAAGAGAGAGCCGTTATCGTCTGTTTGTGGATGTACAGAGTGATATTATTGACACGCCCGGGCGACGGATGGTGATCCCCCTGGCCAGTGCACGTCTGCTGTCAGATAAAGTCTCCCGTGAACTTTACCCGGTGGTGCATATCGGGGATGAAAGCTGGCGCATGATGACCACCGATATGGCCAGTGTGCCGGTCTCCGTTATCGGGGAAGAAGTGGCTGATCTCAGCCACCGCGAAAATGACATCAAAAACGCCATTAACCTGATGTTCTGGGGAATATAAATGTCAGGCTCCCTTATACACAGgcgatgttgaagaccaCGCTGAGGTGTCAATCGTCGGAGCCGCTGAGCAATAACTAGCATAACCCCTTGGGGCCTCTAAACGGGTCTTGAGGGGTTTTTTGCATGGTCATAGCTGTTTCCTGAGAGCTTGGCAGGTGATGACACACATTAACAAATTTCGTGAGGAGTCTCCAGAAGAATGCCATTAATTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAAAATCGACGCTCAAGTCAGAGGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTTCCCCCTGGAAGCTCCCTCGTGCGCTCTCCTGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGGAAGCGTGGCGCTTTCTCATAGCTCACGCTGTAGGTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTGGGCTGTGTGCACGAACCCCCCGTTCAGCCCGACCGCTGCGCCTTATCCGG\", true} // true means it's circular\n","fragment1 := clone.Part{\"GAAGTGCCATTCCGCCTGACCTGAAGACCAGGAGAAACACGTGGCAAACATTCCGGTCTCAAATGGAAAAGAGCAACGAAACCAACGGCTACCTTGACAGCGCTCAAGCCGGCCCTGCAGCTGGCCCGGGCGCTCCGGGTACCGCCGCGGGTCGTGCACGTCGTTGCGCGGGCTTCCTGCGGCGCCAAGCGCTGGTGCTGCTCACGGTGTCTGGTGTTCTGGCAGGCGCCGGTTTGGGCGCGGCACTGCGTGGGCTCAGCCTGAGCCGCACCCAGGTCACCTACCTGGCCTTCCCCGGCGAGATGCTGCTCCGCATGCTGCGCATGATCATCCTGCCGCTGGTGGTCTGCAGCCTGGTGTCGGGCGCCGCCTCCCTCGATGCCAGCTGCCTCGGGCGTCTGGGCGGTATCGCTGTCGCCTACTTTGGCCTCACCACACTGAGTGCCTCGGCGCTCGCCGTGGCCTTGGCGTTCATCATCAAGCCAGGATCCGGTGCGCAGACCCTTCAGTCCAGCGACCTGGGGCTGGAGGACTCGGGGCCTCCTCCTGTCCCCAAAGAAACGGTGGACTCTTTCCTCGACCTGGCCAGAAACCTGTTTCCCTCCAATCTTGTGGTTGCAGCTTTCCGTACGTATGCAACCGATTATAAAGTCGTGACCCAGAACAGCAGCTCTGGAAATGTAACCCATGAAAAGATCCCCATAGGCACTGAGATAGAAGGGATGAACATTTTAGGATTGGTCCTGTTTGCTCTGGTGTTAGGAGTGGCCTTAAAGAAACTAGGCTCCGAAGGAGAGGACCTCATCCGTTTCTTCAATTCCCTCAACGAGGCGACGATGGTGCTGGTGTCCTGGATTATGTGGTACGCGTCTTCAGGCTAGGTGGAGGCTCAGTG\", false}\n","fragment2 := clone.Part{\"GAAGTGCCATTCCGCCTGACCTGAAGACCAGTACGTACCTGTGGGCATCATGTTCCTTGTTGGAAGCAAGATCGTGGAAATGAAAGACATCATCGTGCTGGTGACCAGCCTGGGGAAATACATCTTCGCATCTATATTGGGCCACGTCATTCATGGTGGTATCGTCCTGCCGCTGATTTATTTTGTTTTCACACGAAAAAACCCATTCAGATTCCTCCTGGGCCTCCTCGCCCCATTTGCGACAGCATTTGCTACGTGCTCCAGCTCAGCGACCCTTCCCTCTATGATGAAGTGCATTGAAGAGAACAATGGTGTGGACAAGAGGATCTCCAGGTTTATTCTCCCCATCGGGGCCACCGTGAACATGGACGGAGCAGCCATCTTCCAGTGTGTGGCCGCGGTGTTCATTGCGCAACTCAACAACGTAGAGCTCAACGCAGGACAGATTTTCACCATTCTAGTGACTGCCACAGCGTCCAGTGTTGGAGCAGCAGGCGTGCCAGCTGGAGGGGTCCTCACCATTGCCATTATCCTGGAGGCCATTGGGCTGCCTACTCATGATCTGCCTCTGATCCTGGCTGTGGACTGGATTGTGGACCGGACCACCACGGTGGTGAATGTGGAAGGGGATGCCCTGGGTGCAGGCATTCTCCACCACCTGAATCAGAAGGCAACAAAGAAAGGCGAGCAGGAACTTGCTGAGGTGAAAGTGGAAGCCATCCCCAACTGCAAGTCTGAGGAGGAAACCTCGCCCCTGGTGACACACCAGAACCCCGCTGGCCCCGTGGCCAGTGCCCCAGAACTGGAATCCAAGGAGTCGGTTCTGTGAAGAGCTTAGAGACCGACGACTGCCTAAGGACATTCGCTGCGTCTTCAGGCTAGGTGGAGGCTCAGTG\", false}\n","\n","Clones, _ := clone.GoldenGate([]clone.Part{fragment1, fragment2, popen}, \"BbsI\")\n","Clones"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["[{GGAGAAACACGTGGCAAACATTCCGGTCTCAAATGGAAAAGAGCAACGAAACCAACGGCTACCTTGACAGCGCTCAAGCCGGCCCTGCAGCTGGCCCGGGCGCTCCGGGTACCGCCGCGGGTCGTGCACGTCGTTGCGCGGGCTTCCTGCGGCGCCAAGCGCTGGTGCTGCTCACGGTGTCTGGTGTTCTGGCAGGCGCCGGTTTGGGCGCGGCACTGCGTGGGCTCAGCCTGAGCCGCACCCAGGTCACCTACCTGGCCTTCCCCGGCGAGATGCTGCTCCGCATGCTGCGCATGATCATCCTGCCGCTGGTGGTCTGCAGCCTGGTGTCGGGCGCCGCCTCCCTCGATGCCAGCTGCCTCGGGCGTCTGGGCGGTATCGCTGTCGCCTACTTTGGCCTCACCACACTGAGTGCCTCGGCGCTCGCCGTGGCCTTGGCGTTCATCATCAAGCCAGGATCCGGTGCGCAGACCCTTCAGTCCAGCGACCTGGGGCTGGAGGACTCGGGGCCTCCTCCTGTCCCCAAAGAAACGGTGGACTCTTTCCTCGACCTGGCCAGAAACCTGTTTCCCTCCAATCTTGTGGTTGCAGCTTTCCGTACGTATGCAACCGATTATAAAGTCGTGACCCAGAACAGCAGCTCTGGAAATGTAACCCATGAAAAGATCCCCATAGGCACTGAGATAGAAGGGATGAACATTTTAGGATTGGTCCTGTTTGCTCTGGTGTTAGGAGTGGCCTTAAAGAAACTAGGCTCCGAAGGAGAGGACCTCATCCGTTTCTTCAATTCCCTCAACGAGGCGACGATGGTGCTGGTGTCCTGGATTATGTGGTACGTACCTGTGGGCATCATGTTCCTTGTTGGAAGCAAGATCGTGGAAATGAAAGACATCATCGTGCTGGTGACCAGCCTGGGGAAATACATCTTCGCATCTATATTGGGCCACGTCATTCATGGTGGTATCGTCCTGCCGCTGATTTATTTTGTTTTCACACGAAAAAACCCATTCAGATTCCTCCTGGGCCTCCTCGCCCCATTTGCGACAGCATTTGCTACGTGCTCCAGCTCAGCGACCCTTCCCTCTATGATGAAGTGCATTGAAGAGAACAATGGTGTGGACAAGAGGATCTCCAGGTTTATTCTCCCCATCGGGGCCACCGTGAACATGGACGGAGCAGCCATCTTCCAGTGTGTGGCCGCGGTGTTCATTGCGCAACTCAACAACGTAGAGCTCAACGCAGGACAGATTTTCACCATTCTAGTGACTGCCACAGCGTCCAGTGTTGGAGCAGCAGGCGTGCCAGCTGGAGGGGTCCTCACCATTGCCATTATCCTGGAGGCCATTGGGCTGCCTACTCATGATCTGCCTCTGATCCTGGCTGTGGACTGGATTGTGGACCGGACCACCACGGTGGTGAATGTGGAAGGGGATGCCCTGGGTGCAGGCATTCTCCACCACCTGAATCAGAAGGCAACAAAGAAAGGCGAGCAGGAACTTGCTGAGGTGAAAGTGGAAGCCATCCCCAACTGCAAGTCTGAGGAGGAAACCTCGCCCCTGGTGACACACCAGAACCCCGCTGGCCCCGTGGCCAGTGCCCCAGAACTGGAATCCAAGGAGTCGGTTCTGTGAAGAGCTTAGAGACCGACGACTGCCTAAGGACATTCGCTGAGGTGTCAATCGTCGGAGCCGCTGAGCAATAACTAGCATAACCCCTTGGGGCCTCTAAACGGGTCTTGAGGGGTTTTTTGCATGGTCATAGCTGTTTCCTGAGAGCTTGGCAGGTGATGACACACATTAACAAATTTCGTGAGGAGTCTCCAGAAGAATGCCATTAATTTCCATAGGCTCCGCCCCCCTGACGAGCATCACAAAAATCGACGCTCAAGTCAGAGGTGGCGAAACCCGACAGGACTATAAAGATACCAGGCGTTTCCCCCTGGAAGCTCCCTCGTGCGCTCTCCTGTTCCGACCCTGCCGCTTACCGGATACCTGTCCGCCTTTCTCCCTTCGGGAAGCGTGGCGCTTTCTCATAGCTCACGCTGTAGGTATCTCAGTTCGGTGTAGGTCGTTCGCTCCAAGCTGGGCTGTGTGCACGAACCCCCCGTTCAGCCCGACCGCTGCGCCTTATCCGGTAACTATCGTCTTGAGTCCAACCCGGTAAGACACGACTTATCGCCACTGGCAGCAGCCACTGGTAACAGGATTAGCAGAGCGAGGTATGTAGGCGGTGCTACAGAGTTCTTGAAGTGGTGGCCTAACTACGGCTACACTAGAAGAACAGTATTTGGTATCTGCGCTCTGCTGAAGCCAGTTACCTTCGGAAAAAGAGTTGGTAGCTCTTGATCCGGCAAACAAACCACCGCTGGTAGCGGTGGTTTTTTTGTTTGCAAGCAGCAGATTACGCGCAGAAAAAAAGGATCTCAAGAAGGCCTACTATTAGCAACAACGATCCTTTGATCTTTTCTACGGGGTCTGACGCTCAGTGGAACGAAAACTCACGTTAAGGGATTTTGGTCATGAGATTATCAAAAAGGATCTTCACCTAGATCCTTTTAAATTAAAAATGAAGTTTTAAATCAATCTAAAGTATATATGAGTAAACTTGGTCTGACAGTTACCAATGCTTAATCAGTGAGGCACCTATCTCAGCGATCTGTCTATTTCGTTCATCCATAGTTGCCTGACTCCCCGTCGTGTAGATAACTACGATACGGGAGGGCTTACCATCTGGCCCCAGTGCTGCAATGATACCGCGAGAACCACGCTCACCGGCTCCAGATTTATCAGCAATAAACCAGCCAGCCGGAAGGGCCGAGCGCAGAAGTGGTCCTGCAACTTTATCCGCCTCCATCCAGTCTATTAATTGTTGCCGGGAAGCTAGAGTAAGTAGTTCGCCAGTTAATAGTTTGCGCAACGTTGTTGCCATTGCTACAGGCATCGTGGTGTCACGCTCGTCGTTTGGTATGGCTTCATTCAGCTCCGGTTCCCAACGATCAAGGCGAGTTACATGATCCCCCATGTTGTGCAAAAAAGCGGTTAGCTCCTTCGGTCCTCCGATCGTTGTCAGAAGTAAGTTGGCCGCAGTGTTATCACTCATGGTTATGGCAGCACTGCATAATTCTCTTACTGTCATGCCATCCGTAAGATGCTTTTCTGTGACTGGTGAGTACTCAACCAAGTCATTCTGAGAATAGTGTATGCGGCGACCGAGTTGCTCTTGCCCGGCGTCAATACGGGATAATACCGCGCCACATAGCAGAACTTTAAAAGTGCTCATCATTGGAAAACGTTCTTCGGGGCGAAAACTCTCAAGGATCTTACCGCTGTTGAGATCCAGTTCGATGTAACCCACTCGTGCACCCAACTGATCTTCAGCATCTTTTACTTTCACCAGCGTTTCTGGGTGAGCAAAAACAGGAAGGCAAAATGCCGCAAAAAAGGGAATAAGGGCGACACGGAAATGTTGAATACTCATACTCTTCCTTTTTCAATATTATTGAAGCATTTATCAGGGTTATTGTCTCATGAGCGGATACATATTTGAATGTATTTAGAAAAATAAACAAATAGGGGTTCCGCGCACCTGCACCAGTCAGTAAAACGACGGCCAGTAGTCAAAAGCCTCCGACCGGAGGCTTTTGACTTGGTTCAGGTGGAGTG true}]"]},"metadata":{},"execution_count":12}]},{"cell_type":"markdown","metadata":{"id":"IQ3h2BE-gvb1"},"source":["## Synthesis"]},{"cell_type":"markdown","metadata":{"id":"zYrnuS1Ah7Yf"},"source":["The Synthesis subpackage has one main feature called `fixCDS`. Coding sequences give us the opportunity to change codons to \n","optimize the sequence in a lot of ways like removing undesirables sequences, optimizing codon usage, and lowering GC content. `fixCDS` give us the opportunity to make suggestions of changes and optimize this genetic part."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/","height":79},"id":"5qtwVzATjJZ8","executionInfo":{"status":"error","timestamp":1634755724737,"user_tz":180,"elapsed":98183,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"c35c74f7-3b30-4ef1-827c-fc07cafbbe3d"},"source":["import (\n","  \"github.com/Open-Science-Global/poly/synthesis\"\n","  \"sync\"\n",")\n","\n","seq := \"ATGAGTATTCAACATTTCCGTGTCGCCCTTATTCCCTTTTTTGCGGCATTTTGCCTTCCTGTTTTTGCTCACCCAGAAACGCTGGTGAAAGTAAAAGATGCTGAAGATCAGTTGGGTGCACGAGTGGGTTACATCGAACTGGATCTCAACAGCGGTAAGATCCTTGAGAGTTTTCGCCCCGAAGAACGTTTTCCAATGATGAGCACTTTTAAAGTTCTGCTATGTGGCGCGGTATTATCCCGTATTGACGCCGGGCAAGAGCAACTCGGTCGCCGCATACACTATTCTCAGAATGACTTGGTTGAGTACTCACCAGTCACAGAAAAGCATCTTACGGATGGCATGACAGTAAGAGAATTATGCAGTGCTGCCATAACCATGAGTGATAACACTGCGGCCAACTTACTTCTGACAACGATCGGAGGACCGAAGGAGCTAACCGCTTTTTTGCACAACATGGGGGATCATGTAACTCGCCTTGATCGTTGGGAACCGGAGCTGAATGAAGCCATACCAAACGACGAGCGTGACACCACGATGCCTGTAGCAATGGCAACAACGTTGCGCAAACTATTAACTGGCGAACTACTTACTCTAGCTTCCCGGCAACAATTAATAGACTGGATGGAGGCGGATAAAGTTGCAGGACCACTTCTGCGCTCGGCCCTTCCGGCTGGCTGGTTTATTGCTGATAAATCTGGAGCCGGTGAGCGTGGGTCTCGCGGTATCATTGCAGCACTGGGGCCAGATGGTAAGCCCTCCCGTATCGTAGTTATCTACACGACGGGGAGTCAGGCAACTATGGATGAACGAAATAGACAGATCGCTGAGATAGGTGCCTCACTGATTAAGCATTGGTAA\"\n","sequence := genbank.Read(\"data/ecoli-mg1655.gff\")\n","codonTable := codon.GetCodonTable(11)\n","codingRegions := codon.GetCodingRegions(sequence)\n","optimizationTable := codonTable.OptimizeTable(codingRegions)\n","\n","var functions []func(string, chan synthesis.DnaSuggestion, *sync.WaitGroup)\n","functions = append(functions, synthesis.RemoveSequence([]string{\"GAAGAC\", \"GGTCTC\", \"GCGATG\", \"CGTCTC\", \"GCTCTTC\", \"CACCTGC\"}))\n","\n","fixedSeq, _ := synthesis.FixCds(\":memory:\", seq, optimizationTable, functions)\n","fixedSeq"],"execution_count":null,"outputs":[{"output_type":"error","ename":"ERROR","evalue":"ignored","traceback":["repl.go:15:69: expression returns 3 values, expecting 2: synthesis.FixCds(\":memory:\", bla, optimizationTable, functions)"]}]},{"cell_type":"markdown","metadata":{"id":"-0fC379nj46g"},"source":["## Primer"]},{"cell_type":"markdown","metadata":{"id":"3F2VLXutlD1Q"},"source":["The Primers subpackage includes a \n","- 1) Melting Temperature (Tm) Calculator"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"quaI6dW2lghz","executionInfo":{"status":"ok","timestamp":1626327553411,"user_tz":180,"elapsed":1129,"user":{"displayName":"Isaac Guerreiro","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gibo59sdYK-mXxvtGYnmJyGAQyFx3Ml-_Cizj1www=s64","userId":"12892194237443422762"}},"outputId":"1aa1de8c-cf45-43f7-bd82-193ac2d46440"},"source":["import (\n","  \"fmt\"\n","  \"github.com/Open-Science-Global/poly/primers\"\n",")\n","\n","primer := \"GTAAAACGACGGCCAGT\" //M13 fwd\n","primers.MeltingTemp(primer)"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["52.63382276100299"]},"metadata":{"tags":[]},"execution_count":12}]},{"cell_type":"markdown","metadata":{"id":"DXK2jkHLCtlZ"},"source":["- 2) and a function for creating barcodes if your purpose is sequecing"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"PLdFKvgLmU1U","executionInfo":{"status":"ok","timestamp":1626327556856,"user_tz":180,"elapsed":226,"user":{"displayName":"Isaac Guerreiro","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14Gibo59sdYK-mXxvtGYnmJyGAQyFx3Ml-_Cizj1www=s64","userId":"12892194237443422762"}},"outputId":"e9e1dbf5-c080-4401-d544-5bb911b79a42"},"source":["barcodes := primers.CreateBarcodes(20, 4)\n","barcodes"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["[AAAATAAAGAAACAATTAAT AATGAATCAAGTAAGGAAGC AGCAACTAACGAACCATATA ATAGATACATTTATTGATTC TTCATGTATGGATGCATCTA CTATCGATCCAGAGACAGTT GTTAGTGAGTCAGGTAGGGA GGAGGCAGCTAGCGAGCCAC CACACTTACTGACTCACGTA GTACGGACGCACCTACCGAC GACCCTTTTGTTTCTTGGTT GTTGCTTCGTTCCTGTGTCT TCTGGGTGGCTGCGTGCCTC CTCTCGGTCGCTCCGTCCCG CCGGGGCGGCCGCGCCCCAA]"]},"metadata":{"tags":[]},"execution_count":13}]},{"cell_type":"markdown","metadata":{"id":"OSLnGv8nmokL"},"source":["## Other subpackages"]},{"cell_type":"markdown","metadata":{"id":"jlUPQ1W6mr8N"},"source":["Other subpackages include Random and Seqhash. We highly recommend taking a look to the highly readable and commented **[source code](https://github.com/Open-Science-Global/poly)**, and also the **[documentation](https://pkg.go.dev/github.com/TimothyStiles/poly)** with examples based on real tests.\n","\n","### Several of these subpackages will be looked at in more detail in the following tutorials."]}]}