{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"#4 - FixCDS.ipynb","provenance":[{"file_id":"1NE42_1QHeAyEGTGuAxpsgtGsyJIcIr3I","timestamp":1634684634055},{"file_id":"1JVQfcSFEBoIV4TNCDRYCkcLssQhfp2S2","timestamp":1634426177979},{"file_id":"1GWFFYUcduoykRm08R6Jq7OacGQmDF1BS","timestamp":1634327217511},{"file_id":"1Ts_naBl0P_KfVvwC90Fb-JGWDeP6qjrV","timestamp":1634234790466},{"file_id":"14Bh2CljYULhD03AKCNGlJ3YgM9T5mcRa","timestamp":1627231003334},{"file_id":"1fZTAiBlR5eicaMu4nhhtfeJZEGMUAMHD","timestamp":1626732496856},{"file_id":"1NGDMQi-OGV5-XAbOgz87JWtSc4l2hHyh","timestamp":1626622594804}],"collapsed_sections":["96PZqanZBAle","cZwRFE9UAGIs","3STJpf80Bl9Z"]},"kernelspec":{"display_name":"Go","name":"gophernotes"}},"cells":[{"cell_type":"markdown","metadata":{"id":"dW2UGT4s32-x"},"source":["# **How to: optimize codons with Poly package and friendzymes toolkit**\n","In this notebook, you will input your CDS sequence(s) and receive your CDS corrected without the problematic sequences. This is done by replacing the codons with synonymous ones, thus keeping the same amino acid sequence at the end.\n","\n","*Kind reminder that this tutorial was NOT meant for non-coding sequences such as promoters, rbs, and terminators. If you have found problematic sequences in it, review case by case and be careful not to lose biological meaning.*\n"]},{"cell_type":"markdown","metadata":{"id":"PZJDiu_W3_FT"},"source":["# Configurations for this tutorial\n","\n"]},{"cell_type":"markdown","metadata":{"id":"Xqjf_sf0Am-X"},"source":["\n","First let's run some important settings so you can run this tutorial successfully. \n"]},{"cell_type":"markdown","metadata":{"id":"96PZqanZBAle"},"source":["Colab notebooks use python kernels to run each cell. However, because ***Poly*** is written in **Go language (golang)**, we need to install and configure some things in colab to make feasible run something in go lang here.\n","\n","### **1. In order to start the golang environment, run the line below:**"]},{"cell_type":"code","metadata":{"id":"caT1iMfshw62","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1634764751208,"user_tz":180,"elapsed":96872,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"1a723231-1fee-43ec-fe26-7a9752890974"},"source":["# this process may take a few minutes\n","!add-apt-repository ppa:longsleep/golang-backports -y\n","!apt update\n","!apt install golang-go\n","%env GOPATH=/root/go\n","!go get -u github.com/gopherdata/gophernotes\n","!cp ~/go/bin/gophernotes /usr/bin/\n","!npx degit gopherdata/gophernotes/kernel \\\n","     /usr/local/share/jupyter/kernels/gophernotes"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["\r0% [Working]\r            \rGet:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]\n","Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n","Get:3 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]\n","Get:4 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ Packages [69.5 kB]\n","Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n","Get:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release [696 B]\n","Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n","Get:8 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]\n","Get:9 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release.gpg [836 B]\n","Hit:10 http://archive.ubuntu.com/ubuntu bionic InRelease\n","Get:12 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]\n","Get:13 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Packages [753 kB]\n","Hit:14 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n","Hit:15 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease\n","Get:16 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]\n","Get:17 http://security.ubuntu.com/ubuntu bionic-security/universe amd64 Packages [1,434 kB]\n","Hit:18 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n","Get:19 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [2,213 kB]\n","Get:20 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic InRelease [15.4 kB]\n","Get:21 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [2,385 kB]\n","Get:22 http://security.ubuntu.com/ubuntu bionic-security/restricted amd64 Packages [626 kB]\n","Get:23 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main Sources [1,809 kB]\n","Get:24 http://archive.ubuntu.com/ubuntu bionic-updates/restricted amd64 Packages [664 kB]\n","Get:25 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [2,832 kB]\n","Get:26 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main amd64 Packages [926 kB]\n","Get:27 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 Packages [6,639 B]\n","Fetched 14.0 MB in 7s (1,961 kB/s)\n","Reading package lists... Done\n","Hit:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease\n","Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease\n","Hit:3 http://security.ubuntu.com/ubuntu bionic-security InRelease\n","Ign:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease\n","Hit:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release\n","Hit:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release\n","Hit:7 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease\n","Hit:8 http://archive.ubuntu.com/ubuntu bionic InRelease\n","Hit:10 http://archive.ubuntu.com/ubuntu bionic-updates InRelease\n","Hit:11 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease\n","Hit:13 http://archive.ubuntu.com/ubuntu bionic-backports InRelease\n","Hit:14 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease\n","Hit:15 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease\n","Hit:16 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic InRelease\n","Reading package lists... Done\n","Building dependency tree       \n","Reading state information... Done\n","67 packages can be upgraded. Run 'apt list --upgradable' to see them.\n","Reading package lists... Done\n","Building dependency tree       \n","Reading state information... Done\n","The following additional packages will be installed:\n","  golang-1.17-go golang-1.17-src golang-src\n","Suggested packages:\n","  bzr | brz mercurial subversion\n","The following NEW packages will be installed:\n","  golang-1.17-go golang-1.17-src golang-go golang-src\n","0 upgraded, 4 newly installed, 0 to remove and 67 not upgraded.\n","Need to get 71.8 MB of archives.\n","After this operation, 423 MB of additional disk space will be used.\n","Get:1 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-1.17-src all 1.17.2-1longsleep1+focal [15.2 MB]\n","Get:2 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-1.17-go amd64 1.17.2-1longsleep1+focal [56.6 MB]\n","Get:3 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-src amd64 2:1.17~1longsleep1 [5,056 B]\n","Get:4 http://ppa.launchpad.net/longsleep/golang-backports/ubuntu bionic/main amd64 golang-go amd64 2:1.17~1longsleep1 [24.1 kB]\n","Fetched 71.8 MB in 11s (6,306 kB/s)\n","Selecting previously unselected package golang-1.17-src.\n","(Reading database ... 155047 files and directories currently installed.)\n","Preparing to unpack .../golang-1.17-src_1.17.2-1longsleep1+focal_all.deb ...\n","Unpacking golang-1.17-src (1.17.2-1longsleep1+focal) ...\n","Selecting previously unselected package golang-1.17-go.\n","Preparing to unpack .../golang-1.17-go_1.17.2-1longsleep1+focal_amd64.deb ...\n","Unpacking golang-1.17-go (1.17.2-1longsleep1+focal) ...\n","Selecting previously unselected package golang-src:amd64.\n","Preparing to unpack .../golang-src_2%3a1.17~1longsleep1_amd64.deb ...\n","Unpacking golang-src:amd64 (2:1.17~1longsleep1) ...\n","Selecting previously unselected package golang-go.\n","Preparing to unpack .../golang-go_2%3a1.17~1longsleep1_amd64.deb ...\n","Unpacking golang-go (2:1.17~1longsleep1) ...\n","Setting up golang-1.17-src (1.17.2-1longsleep1+focal) ...\n","Setting up golang-src:amd64 (2:1.17~1longsleep1) ...\n","Setting up golang-1.17-go (1.17.2-1longsleep1+focal) ...\n","Setting up golang-go (2:1.17~1longsleep1) ...\n","Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n","env: GOPATH=/root/go\n","go: downloading github.com/gopherdata/gophernotes v0.7.3\n","go: downloading github.com/cosmos72/gomacro v0.0.0-20210624153544-b4935e406a41\n","go: downloading github.com/go-zeromq/zmq4 v0.13.0\n","go: downloading github.com/gofrs/uuid v4.0.0+incompatible\n","go: downloading github.com/gofrs/uuid v4.1.0+incompatible\n","go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1\n","go: downloading github.com/peterh/liner v1.2.1\n","go: downloading github.com/go-zeromq/goczmq/v4 v4.2.2\n","go: downloading golang.org/x/sync v0.0.0-20201020160332-67f06af15bc9\n","go: downloading github.com/go-zeromq/goczmq v4.1.0+incompatible\n","go: downloading golang.org/x/sync v0.0.0-20210220032951-036812b2e83c\n","go: downloading golang.org/x/tools v0.1.0\n","go: downloading github.com/mattn/go-runewidth v0.0.12\n","go: downloading github.com/mattn/go-runewidth v0.0.13\n","go: downloading golang.org/x/tools v0.1.7\n","go: downloading github.com/rivo/uniseg v0.1.0\n","go: downloading golang.org/x/sys v0.0.0-20210119212857-b64e53b001e4\n","go: downloading github.com/rivo/uniseg v0.2.0\n","go: downloading golang.org/x/mod v0.3.0\n","go: downloading golang.org/x/mod v0.5.1\n","go: downloading golang.org/x/sys v0.0.0-20211020174200-9d6173849985\n","go get: installing executables with 'go get' in module mode is deprecated.\n","\tUse 'go install pkg@version' instead.\n","\tFor more information, see https://golang.org/doc/go-get-install-deprecation\n","\tor run 'go help get' or 'go help install'.\n","\u001b[K\u001b[?25hnpx: installed 1 in 2.494s\n","\u001b[36m> cloned \u001b[1mgopherdata/gophernotes\u001b[22m#\u001b[1mHEAD\u001b[22m to /usr/local/share/jupyter/kernels/gophernotes\u001b[39m\n"]}]},{"cell_type":"markdown","metadata":{"id":"cZwRFE9UAGIs"},"source":["### **2. Download important data to run this tutorial**"]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"wHmBJ418_5tX","executionInfo":{"status":"ok","timestamp":1634764764895,"user_tz":180,"elapsed":531,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"2d36e160-4d3d-4a2d-d3b9-f247cb16e53f"},"source":["!rm -rf $GOPATH/pkg/mod/github.com/!open-!science-!global\n","!rm -rf $GOPATH/pkg/mod/cache/download/github.com/!open-!science-!global\n","!go get -u github.com/Open-Science-Global/poly@e3e1c61"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["go: downloading github.com/Open-Science-Global/poly v0.11.3\n"]}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"uaxY1xbR-U0F","executionInfo":{"status":"ok","timestamp":1634764770413,"user_tz":180,"elapsed":1082,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"402928ba-9a96-4b32-9fd6-9b497e3039bb"},"source":["!wget https://raw.githubusercontent.com/Open-Science-Global/friendzymes-toolkit/main/data/codon-table/bsub-ecoli.json"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["--2021-10-20 21:19:29--  https://raw.githubusercontent.com/Open-Science-Global/friendzymes-toolkit/main/data/codon-table/bsub-ecoli.json\n","Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...\n","Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 4702 (4.6K) [text/plain]\n","Saving to: ‘bsub-ecoli.json’\n","\n","bsub-ecoli.json     100%[===================>]   4.59K  --.-KB/s    in 0s      \n","\n","2021-10-20 21:19:29 (47.1 MB/s) - ‘bsub-ecoli.json’ saved [4702/4702]\n","\n"]}]},{"cell_type":"markdown","metadata":{"id":"3STJpf80Bl9Z"},"source":["### **3. Connect Colab Notebook to your GDrive (not required)**"]},{"cell_type":"markdown","metadata":{"id":"ftGzhtH5AOZM"},"source":["The previous code will download the files temporarily. If you want to download them to a folder on your drive and save it for later analysis or if you are already using this notebook to run your own files, you should connect your Google Drive to Colab and you will be able to access, read and save files permanently. \n","So, if you prefer, you can do this with the code line below:\n","\n"]},{"cell_type":"code","metadata":{"id":"sh661uGn_8R1"},"source":["from google.colab import drive\n","drive.mount('/content/drive')"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"GbPLWTJ05uH4"},"source":["## After running these steps, click on **Runtime** in the menu bar & **Change Runtime Type** to Go, if it hasn't changed yet.\n","\n","This will make Colab starting use a open source go kernel called Gopher Notes.\n","\n","Now we are ready to work."]},{"cell_type":"markdown","metadata":{"id":"34gduXzXl-Nb"},"source":["# **Importing packages and pre-requisites**"]},{"cell_type":"code","metadata":{"id":"gQVnSh-tmEvz"},"source":["package main\n","\n","import (\n","\t\"fmt\"\n","\t\"os\"\n","\t\"path/filepath\"\n","\t\"strings\"\n","\t\"sync\"\n","\n","  \"github.com/Open-Science-Global/poly\"\n","\t\"github.com/Open-Science-Global/poly/checks\"\n","\t\"github.com/Open-Science-Global/poly/io/fasta\"\n","\t\"github.com/Open-Science-Global/poly/io/genbank\"\n","\t\"github.com/Open-Science-Global/poly/linearfold\"\n","\t\"github.com/Open-Science-Global/poly/synthesis\"\n","\t\"github.com/Open-Science-Global/poly/transform\"\n","\t\"github.com/Open-Science-Global/poly/transform/codon\"\n",")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Ci4PRegmP-pd"},"source":["## **Fixing problematic sequences**\n","\n","Keep in mind that the sequence you give as input for codon optimization has to be:\n","\n","1.   a coding sequence (CDS);\n","2.   in a DNA sequence format, i.e. in A, T, C and Gs;\n"]},{"cell_type":"code","metadata":{"id":"uJqfzQ90qY-n"},"source":["// run this to set up the subfunction 1 (REMOVE OR ADD RESTRICTIONS ENZYMES BINDING SITES AS YOU WISH)\n","\n","func forbiddenSeqList() []string {\n","\tBsaI := \"GGTCTC\"\n","\tBbsI := \"GAAGAC\"\n","\tBtgzI := \"GCGATG\"\n","\tSapI := \"GCTCTTC\"\n","\tBsmbI := \"CGTCTC\"\n","\tAarI := \"CACCTGC\"\n","\tPmeI := \"GTTTAAAC\"\n","\tHindIII := \"AAGCTT\"\n","\tPstI := \"CTGCAG\"\n","\tXbaI := \"TCTAGA\"\n","\tBamHI := \"GGATCC\"\n","\tSmaI := \"CCCGGG\"\n","\tKpnI := \"GGTACC\"\n","\tSacI := \"GAGCTC\"\n","\tSalI := \"GTCGAC\"\n","\tEcoRI := \"GAATTC\"\n","\tSphI := \"GCATGC\"\n","\tAvrII := \"CCTAGG\"\n","\tSwaI := \"ATTTAAAT\"\n","\tAscI := \"GGCGCGCC\"\n","\tFseI := \"GGCCGGCC\"\n","\tPacI := \"TTAATTAA\"\n","\tSpeI := \"ACTAGT\"\n","\tNotI := \"GCGGCCGC\"\n","\tSanDI_A := \"GGGACCC\"\n","\tSanDI_T := \"GGGTCCC\"\n","\tBglII := \"AGATCT\"\n","\tXhoI := \"CTCGAG\"\n","\tClaI := \"ATCGAT\"\n","\n","\tList := []string{\n","\t\tBsaI,\n","\t\tBbsI,\n","\t\tSapI,\n","\t\tBsmbI,\n","\t\tBtgzI,\n","\t\tAarI,\n","\t\tPmeI,\n","\t\tHindIII,\n","\t\tPstI,\n","\t\tXbaI,\n","\t\tBamHI,\n","\t\tSmaI,\n","\t\tKpnI,\n","\t\tSacI,\n","\t\tSalI,\n","\t\tEcoRI,\n","\t\tSphI,\n","\t\tAvrII,\n","\t\tSacI,\n","\t\tSalI,\n","\t\tSwaI,\n","\t\tAscI,\n","\t\tFseI,\n","\t\tPacI,\n","\t\tSpeI,\n","\t\tNotI,\n","\t\tSanDI_A,\n","\t\tSanDI_T,\n","\t\tBglII,\n","\t\tXhoI,\n","\t\tClaI,\n","\t}\n","\n","\treturn List\n","}"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"2l-5ddmsqZt0"},"source":["// run this to set up the subfunction 2\n","\n","func homologySequencesFindProblems() []string {\n","\treturn []string{\"AAAAAA\", \"CCCCCC\"}\n","}"],"execution_count":null,"outputs":[]},{"cell_type":"code","metadata":{"id":"02c_x8KRgqkQ"},"source":["// run this to set up the main function\n","\n","func fixSequence(sequence string, codonTable codon.Table) string {\n","\t// Construct function that will remove unwanted properties in our sequences\n","\t// Function#1: Remove unwanted sequences as restriction binding sites and homopolymers of length 5\n","\tforbiddenSequences := forbiddenSeqList()\n","\tremoveSequenceFunc := synthesis.RemoveSequence(forbiddenSequences)\n","\n","\t// Function#2: Remove secondary structures\n","\tremoveSecondaryFunc := synthesis.RemoveHairpin(20, 200)\n","\n","\t// Function#3: Remove repetition greater than 10 inside the sequence\n","\tremoveRepeatFunc := synthesis.RemoveRepeat(10)\n","\n","\t// Added all those functions to a list and pass through FixCds function that will take care of our problems\n","\tvar functions []func(string, chan synthesis.DnaSuggestion, *sync.WaitGroup)\n","\tfunctions = append(functions, removeSequenceFunc, removeRepeatFunc, removeSecondaryFunc)\n","\n","\tfixedSeq, _, _ := synthesis.FixCds(\":memory:\", sequence, codonTable, functions)\n","\t// Because FixCds actually remove stop codon we will concatenate to it\n","\treturn fixedSeq\n","}\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"0hCCKtmwh9Zd"},"source":["The codon table is very important for this step, because it tells us which synonymous codon to use in case we have problematic parts that need correction. \n","\n","For this tutorial, we have already downloaded this data (`\"bsub-ecoli.json\"`) in the settings part (in \"Download important data to run this tutorial\"). In previous tutorials, like Codon Optimization, we have how to generate this codon table with the function `GenerateCodonTable()`, in case you want to generate your own."]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pMHh0U-FJ44J","executionInfo":{"status":"ok","timestamp":1634765015974,"user_tz":180,"elapsed":575,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"742e0de1-fecd-406a-e413-15ee95f321df"},"source":["codonTable := codon.ReadCodonJSON(\"bsub-ecoli.json\")\n","fmt.Println(codonTable)"],"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["{[TTG CTG ATT ATC ATA ATG GTG] [TAA TAG TGA] [{Q [{CAA 4471} {CAG 5528}]} {R [{CGT 1496} {CGC 2527} {CGA 1619} {CGG 2037} {AGA 1122} {AGG 1196}]} {N [{AAT 4311} {AAC 5688}]} {A [{GCT 2078} {GCC 2117} {GCA 2392} {GCG 3411}]} {E [{GAA 7067} {GAG 2932}]} {Y [{TAT 5296} {TAC 4703}]} {* [{TAA 2820} {TAG 1197} {TGA 5982}]} {T [{ACT 1952} {ACC 2695} {ACA 2266} {ACG 3085}]} {S [{TCT 1498} {TCC 1300} {TCA 1910} {TCG 2130} {AGT 1150} {AGC 2008}]} {C [{TGT 3531} {TGC 6468}]} {I [{ATT 3513} {ATC 3786} {ATA 2700}]} {M [{ATG 10000}]} {K [{AAA 5743} {AAG 4256}]} {D [{GAT 6146} {GAC 3853}]} {G [{GGT 2480} {GGC 3515} {GGA 2052} {GGG 1951}]} {P [{CCT 1914} {CCC 1597} {CCA 2618} {CCG 3870}]} {H [{CAT 4930} {CAC 5069}]} {W [{TGG 10000}]} {V [{GTT 2662} {GTC 2133} {GTA 2085} {GTG 3118}]} {F [{TTT 5416} {TTC 4583}]} {L [{TTA 1873} {TTG 2390} {CTT 1100} {CTC 0} {CTA 0} {CTG 2885}]}]}\n"]},{"output_type":"execute_result","data":{"text/plain":["874 <nil>"]},"metadata":{},"execution_count":5}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ixsnxj1zh9xv","executionInfo":{"status":"ok","timestamp":1634765130092,"user_tz":180,"elapsed":413,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"52300e7b-1e8f-4ed6-bdc2-42b45c0ed000"},"source":["// run this line to read the CDS example to fix\n","// For this tutorial we will use the DNA sequence that corresponds to the enzyme Pfu-Sso7d, \n","// which was put into the variable below. This sequence has 3420 base pairs.\n","\n","PfuSeq := \"ATGGGTCTCATTTTAGATGTGGATTATATCACAGAGGAAGGAAAGCCAGTTATACGTCTTTTCAAGAAGGAAAATGGGAAATTTAAGATAGAGCATGACCGTACATTCCGTCCGTATATCTATGCCCTGTTGCGTGATGATTCTAAAATCGAGGAAGTCAAGAAAATTACCGGCGAACGGCACGGTAAAATAGTCCGGATCGTGGACGTAGAAAAGGTAGAAAAGAAATTCCTGGGGAAACCGATAACTGTATGGAAGCTGTATCTTGAACATCCGCAAGACGTCCCAACTATTCGAGAGAAAGTTAGAGAACATCCGGCAGTGGTGGATATTTTCGAATATGACATACCGTTTGCCAAACGATATTTGATAGACAAAGGTCTGATCCCGATGGAAGGGGAAGAGGAGCTGAAAATTTTGGCGTTTGATATCGAAACTTTGTATCATGAAGGCGAAGAATTCGGTAAGGGCCCTATCATCATGATCAGTTATGCAGATGAAAACGAAGCTAAGGTGATTACGTGGAAGAACATAGATTTGCCTTATGTCGAGGTAGTGTCATCAGAGCGTGAGATGATCAAGCGCTTCTTGCGTATTATTCGTGAGAAAGATCCTGACATCATTGTTACCTATAATGGAGATTCATTTGACTTTCCTTATTTAGCTAAGCGTGCCGAGAAATTAGGCATTAAGCTTACCATTGGGAGAGATGGATCGGAACCGAAAATGCAGAGGATTGGCGACATGACGGCAGTTGAAGTGAAAGGAAGAATCCACTTTGACTTGTATCACGTGATAACGAGAACAATCAATCTGCCCACTTACACACTAGAAGCTGTTTATGAAGCCATATTTGGAAAACCCAAGGAAAAGGTATACGCAGACGAGATCGCGAAGGCGTGGGAGTCTGGTGAAAACTTAGAGAGGGTTGCAAAGTATTCTATGGAGGATGCAAAAGCTACGTACGAATTGGGTAAAGAGTTCTTGCCCATGGAGATACAGTTGTCCCGGCTGGTGGGACAACCCCTGTGGGATGTGTCGCGCTCGTCAACTGGCAACTTAGTTGAATGGTTTCTTTTGCGGAAGGCGTATGAACGCAACGAAGTCGCGCCAAATAAGCCAAGCGAGGAGGAATATCAAAGACGATTGAGGGAAAGTTACACCGGTGGCTTCGTTAAAGAACCTGAGAAAGGGTTGTGGGAAAACATCGTATACCTGGACTTCAGGGCATTATACCCTAGCATCATCATTACGCACAACGTGAGCCCCGATACATTGAATCTTGAGGGGTGCAAGAATTACGATATCGCCCCGCAGGTAGGCCATAAATTTTGCAAAGATATACCGGGCTTCATACCATCACTTTTAGGACACCTGTTAGAAGAACGGCAAAAGATTAAAACAAAGATGAAGGAAACACAGGACCCGATAGAGAAGATTCTGTTAGATTACCGGCAAAAGGCAATCAAACTTTTAGCAAACAGCTTTTACGGGTACTACGGGTACGCAAAAGCACGATGGTATTGCAAAGAATGTGCAGAATCAGTCACTGCATGGGGTCGTAAGTATATAGAATTGGTCTGGAAGGAATTAGAAGAGAAGTTCGGCTTTAAAGTGTTATACATAGACACGGATGGTCTTTATGCCACTATTCCTGGTGGAGAAAGTGAAGAGATTAAGAAGAAAGCATTAGAGTTCGTTAAGTATATCAACTCGAAATTGCCCGGTCTGCTGGAACTGGAATACGAAGGTTTCTATAAGCGCGGATTCTTCGTAACGAAGAAGCGCTATGCGGTCATCGATGAAGAGGGAAAAGTTATTACGCGCGGTCTGGAGATAGTCCGTAGGGACTGGTCGGAAATCGCGAAAGAAACACAAGCGCGAGTGTTGGAAACCATCTTAAAGCACGGGGACGTTGAGGAGGCTGTTAGGATTGTAAAAGAAGTCATCCAGAAGCTGGCAAATTATGAAATCCCTCCTGAGAAACTGGCCATCTACGAGCAAATTACTAGACCCTTGCATGAGTATAAGGCAATTGGTCCACATGTGGCGGTGGCTAAGAAACTTGCGGCGAAGGGTGTTAAAATCAAACCCGGCATGGTAATTGGATATATTGTGCTGAGGGGAGATGGTCCTATTTCCAATCGCGCCATATTAGCAGAGGAATATGATCCAAAGAAGCACAAATACGATGCTGAATACTATATCGAAAATCAGGTCTTACCCGCGGTATTGCGTATATTAGAGGGCTTTGGGTACCGCAAAGAGGACTTGCGGTATCAGAAAACGAGACAGGTGGGATTGACGTCTTGGTTAAATATAAAGAAGTCGGGGACGGGCGGAGGCGGCGCAACCGTAAAATTTAAGTACAAAGGCGAGGAGAAAGAAGTTGACATCAGCAAAATCAAGAAAGTGTGGCGCGTTGGGAAAATGATCTCGTTCACTTATGATGAAGGTGGCGGCAAAACCGGTCGTGGCGCCGTATCGGAGAAAGATGCACCAAAGGAACTTTTACAGATGTTAGAGAAACAGAAGAAAGGAGGTGGCTCAGGCGGCGGATCGGAAAACCTGTATTTTCAGGGTGGAGGCGGGTCCATGGTATCAAGTGGAGAAGATATTTTCTCCGGCCTGGTGCCTATCCTGATTGAGTTAGAGGGCGACGTTAATGGCCACAGATTTTCGGTTCGTGGCGAAGGATATGGTGATGCTTCAAATGGCAAATTGGAGATCAAATTTATTTGCACCACGGGTCGCTTGCCGGTCCCGTGGCCGACCCTGGTAACAACGCTGAGTTATGGAGTCCAGTGCTTTGCGAAATACCCAGAGCACATGCGCCAGAATGACTTCTTTAAGAGCGCAATGCCGGATGGCTATGTTCAGGAGCGCACGATCAGCTTTAAAGAGGACGGTACCTACAAAACTCGAGCAGAAGTTAAATTCGAAGGGGAGGCGCTTGTTAACCGGATCGACCTGAAGGGCTTAGAATTCAAAGAGGATGGAAACATCTTGGGTCACAAACTTGAATACTCATTTAATAGTCACTATGTGTATATTACGGCGGATAAGAACAGGAATGGACTGGAGGCACAGTTTCGAATCCGGCATAATGTCGATGATGGTTCCGTCCAATTGGCGGACCATTACCAGCAAAATACCCCGATCGGGGAGGGCCCCGTGTTGTTACCTGAACAGCACTATTTAACCACCAATAGCGTTCTTTCTAAAGATCCCCAAGAACGCAGGGACCACATGGTTCTAGTTGAATTTGTGACTGCTGCCGGTTTGAGCTTAGGAATGGATGAATTGTACAAAAGCGGTGGAGGCAGCCACCATCATCATCATCATCATCACCATCACTCTTCCAAGAAATCAGGTTCATACTCTGGTTCCAAAGGCTCAAAACGGCGGATTCTGTAATAA\"\n","PfuSeq"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["ATGGGTCTCATTTTAGATGTGGATTATATCACAGAGGAAGGAAAGCCAGTTATACGTCTTTTCAAGAAGGAAAATGGGAAATTTAAGATAGAGCATGACCGTACATTCCGTCCGTATATCTATGCCCTGTTGCGTGATGATTCTAAAATCGAGGAAGTCAAGAAAATTACCGGCGAACGGCACGGTAAAATAGTCCGGATCGTGGACGTAGAAAAGGTAGAAAAGAAATTCCTGGGGAAACCGATAACTGTATGGAAGCTGTATCTTGAACATCCGCAAGACGTCCCAACTATTCGAGAGAAAGTTAGAGAACATCCGGCAGTGGTGGATATTTTCGAATATGACATACCGTTTGCCAAACGATATTTGATAGACAAAGGTCTGATCCCGATGGAAGGGGAAGAGGAGCTGAAAATTTTGGCGTTTGATATCGAAACTTTGTATCATGAAGGCGAAGAATTCGGTAAGGGCCCTATCATCATGATCAGTTATGCAGATGAAAACGAAGCTAAGGTGATTACGTGGAAGAACATAGATTTGCCTTATGTCGAGGTAGTGTCATCAGAGCGTGAGATGATCAAGCGCTTCTTGCGTATTATTCGTGAGAAAGATCCTGACATCATTGTTACCTATAATGGAGATTCATTTGACTTTCCTTATTTAGCTAAGCGTGCCGAGAAATTAGGCATTAAGCTTACCATTGGGAGAGATGGATCGGAACCGAAAATGCAGAGGATTGGCGACATGACGGCAGTTGAAGTGAAAGGAAGAATCCACTTTGACTTGTATCACGTGATAACGAGAACAATCAATCTGCCCACTTACACACTAGAAGCTGTTTATGAAGCCATATTTGGAAAACCCAAGGAAAAGGTATACGCAGACGAGATCGCGAAGGCGTGGGAGTCTGGTGAAAACTTAGAGAGGGTTGCAAAGTATTCTATGGAGGATGCAAAAGCTACGTACGAATTGGGTAAAGAGTTCTTGCCCATGGAGATACAGTTGTCCCGGCTGGTGGGACAACCCCTGTGGGATGTGTCGCGCTCGTCAACTGGCAACTTAGTTGAATGGTTTCTTTTGCGGAAGGCGTATGAACGCAACGAAGTCGCGCCAAATAAGCCAAGCGAGGAGGAATATCAAAGACGATTGAGGGAAAGTTACACCGGTGGCTTCGTTAAAGAACCTGAGAAAGGGTTGTGGGAAAACATCGTATACCTGGACTTCAGGGCATTATACCCTAGCATCATCATTACGCACAACGTGAGCCCCGATACATTGAATCTTGAGGGGTGCAAGAATTACGATATCGCCCCGCAGGTAGGCCATAAATTTTGCAAAGATATACCGGGCTTCATACCATCACTTTTAGGACACCTGTTAGAAGAACGGCAAAAGATTAAAACAAAGATGAAGGAAACACAGGACCCGATAGAGAAGATTCTGTTAGATTACCGGCAAAAGGCAATCAAACTTTTAGCAAACAGCTTTTACGGGTACTACGGGTACGCAAAAGCACGATGGTATTGCAAAGAATGTGCAGAATCAGTCACTGCATGGGGTCGTAAGTATATAGAATTGGTCTGGAAGGAATTAGAAGAGAAGTTCGGCTTTAAAGTGTTATACATAGACACGGATGGTCTTTATGCCACTATTCCTGGTGGAGAAAGTGAAGAGATTAAGAAGAAAGCATTAGAGTTCGTTAAGTATATCAACTCGAAATTGCCCGGTCTGCTGGAACTGGAATACGAAGGTTTCTATAAGCGCGGATTCTTCGTAACGAAGAAGCGCTATGCGGTCATCGATGAAGAGGGAAAAGTTATTACGCGCGGTCTGGAGATAGTCCGTAGGGACTGGTCGGAAATCGCGAAAGAAACACAAGCGCGAGTGTTGGAAACCATCTTAAAGCACGGGGACGTTGAGGAGGCTGTTAGGATTGTAAAAGAAGTCATCCAGAAGCTGGCAAATTATGAAATCCCTCCTGAGAAACTGGCCATCTACGAGCAAATTACTAGACCCTTGCATGAGTATAAGGCAATTGGTCCACATGTGGCGGTGGCTAAGAAACTTGCGGCGAAGGGTGTTAAAATCAAACCCGGCATGGTAATTGGATATATTGTGCTGAGGGGAGATGGTCCTATTTCCAATCGCGCCATATTAGCAGAGGAATATGATCCAAAGAAGCACAAATACGATGCTGAATACTATATCGAAAATCAGGTCTTACCCGCGGTATTGCGTATATTAGAGGGCTTTGGGTACCGCAAAGAGGACTTGCGGTATCAGAAAACGAGACAGGTGGGATTGACGTCTTGGTTAAATATAAAGAAGTCGGGGACGGGCGGAGGCGGCGCAACCGTAAAATTTAAGTACAAAGGCGAGGAGAAAGAAGTTGACATCAGCAAAATCAAGAAAGTGTGGCGCGTTGGGAAAATGATCTCGTTCACTTATGATGAAGGTGGCGGCAAAACCGGTCGTGGCGCCGTATCGGAGAAAGATGCACCAAAGGAACTTTTACAGATGTTAGAGAAACAGAAGAAAGGAGGTGGCTCAGGCGGCGGATCGGAAAACCTGTATTTTCAGGGTGGAGGCGGGTCCATGGTATCAAGTGGAGAAGATATTTTCTCCGGCCTGGTGCCTATCCTGATTGAGTTAGAGGGCGACGTTAATGGCCACAGATTTTCGGTTCGTGGCGAAGGATATGGTGATGCTTCAAATGGCAAATTGGAGATCAAATTTATTTGCACCACGGGTCGCTTGCCGGTCCCGTGGCCGACCCTGGTAACAACGCTGAGTTATGGAGTCCAGTGCTTTGCGAAATACCCAGAGCACATGCGCCAGAATGACTTCTTTAAGAGCGCAATGCCGGATGGCTATGTTCAGGAGCGCACGATCAGCTTTAAAGAGGACGGTACCTACAAAACTCGAGCAGAAGTTAAATTCGAAGGGGAGGCGCTTGTTAACCGGATCGACCTGAAGGGCTTAGAATTCAAAGAGGATGGAAACATCTTGGGTCACAAACTTGAATACTCATTTAATAGTCACTATGTGTATATTACGGCGGATAAGAACAGGAATGGACTGGAGGCACAGTTTCGAATCCGGCATAATGTCGATGATGGTTCCGTCCAATTGGCGGACCATTACCAGCAAAATACCCCGATCGGGGAGGGCCCCGTGTTGTTACCTGAACAGCACTATTTAACCACCAATAGCGTTCTTTCTAAAGATCCCCAAGAACGCAGGGACCACATGGTTCTAGTTGAATTTGTGACTGCTGCCGGTTTGAGCTTAGGAATGGATGAATTGTACAAAAGCGGTGGAGGCAGCCACCATCATCATCATCATCATCACCATCACTCTTCCAAGAAATCAGGTTCATACTCTGGTTCCAAAGGCTCAAAACGGCGGATTCTGTAATAA"]},"metadata":{},"execution_count":8}]},{"cell_type":"code","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"JhvPYmKRqjh7","executionInfo":{"status":"ok","timestamp":1634765143649,"user_tz":180,"elapsed":467,"user":{"displayName":"Giovanna Maklouf","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GiwcMODmckyDjtciji3QIpNCZ6Keh-_crpfnMyIBg=s64","userId":"02331481017878350508"}},"outputId":"166cc8a3-f621-4127-c518-255c9e967d63"},"source":["// run this line to actually fix the sequence\n","\n","fixedSequence := fixSequence(PfuSeq, codonTable)\n","fixedSequence"],"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["ATGGGACTCATTTTAGATGTGGATTATATCACAGAGGAAGGAAAGCCAGTTATACGTCTTTTCAAGAAGGAAAATGGGAAATTTAAGATAGAGCATGACCGTACATTCCGTCCGTATATCTATGCCCTGTTGCGTGATGATTCTAAAATCGAGGAAGTCAAGAAAATTACCGGCGAACGGCACGGTAAAATAGTCCGGATCGTGGACGTAGAAAAGGTAGAAAAGAAATTCCTGGGGAAACCGATAACTGTATGGAAGCTGTATCTTGAACATCCGCAAGACGTCCCAACTATTCGAGAGAAAGTTAGAGAACATCCGGCAGTGGTGGATATTTTCGAATATGACATACCGTTTGCCAAACGATATTTGATAGACAAAGGTCTGATCCCGATGGAAGGGGAAGAGGAGCTGAAAATTTTGGCGTTTGATATCGAAACTTTGTATCATGAAGGCGAAGAGTTCGGTAAGGGCCCTATCATCATGATCAGTTATGCAGATGAAAACGAAGCTAAGGTGATTACGTGGAAGAACATAGATTTGCCTTATGTCGAGGTAGTGTCATCAGAGCGTGAGATGATCAAGCGCTTCTTGCGTATTATTCGTGAGAAAGATCCTGACATCATTGTTACCTATAATGGAGATTCATTTGACTTTCCTTATTTAGCTAAGCGTGCCGAGAAATTAGGCATTAAACTTACCATTGGGAGAGATGGATCGGAACCGAAAATGCAGAGGATTGGCGACATGACGGCAGTTGAAGTGAAAGGAAGAATCCACTTTGACTTGTATCACGTGATAACGAGAACAATCAATCTGCCCACTTACACACTAGAAGCTGTTTATGAAGCCATATTTGGAAAACCCAAGGAAAAGGTATACGCAGACGAGATCGCGAAGGCGTGGGAGTCTGGTGAAAACTTAGAGAGGGTTGCAAAGTATTCTATGGAGGATGCAAAAGCTACGTACGAATTGGGTAAAGAGTTCTTGCCCATGGAGATACAGTTGTCCCGGCTGGTGGGACAACCCCTGTGGGATGTGTCGCGCTCGTCAACTGGCAACTTAGTTGAATGGTTTCTTTTGCGGAAGGCGTATGAACGCAACGAAGTCGCGCCAAATAAGCCAAGCGAGGAGGAATATCAAAGACGATTGAGGGAAAGTTACACCGGTGGCTTCGTTAAAGAACCTGAGAAAGGGTTGTGGGAAAACATCGTATACCTGGACTTCAGGGCATTATACCCTAGCATCATCATTACGCACAACGTGAGCCCCGATACATTGAATCTTGAGGGGTGCAAGAATTACGATATCGCCCCGCAGGTAGGCCATAAATTTTGCAAAGATATACCGGGCTTCATACCATCACTTTTAGGACACCTGTTAGAAGAACGGCAAAAGATTAAAACAAAGATGAAGGAAACACAGGACCCGATAGAGAAGATTCTGTTAGATTACCGGCAAAAGGCAATCAAACTTTTAGCAAACAGCTTTTACGGGTACTACGGGTACGCAAAAGCACGATGGTATTGCAAAGAATGTGCAGAATCAGTCACTGCATGGGGTCGTAAGTATATAGAATTGGTCTGGAAGGAATTAGAAGAGAAGTTCGGCTTTAAAGTGTTATACATAGACACGGATGGTCTTTATGCCACTATTCCTGGTGGAGAAAGTGAAGAGATTAAGAAGAAAGCATTAGAGTTCGTTAAGTATATCAACTCGAAATTGCCCGGTCTGCTGGAACTGGAATACGAAGGTTTCTATAAGCGCGGATTCTTCGTAACGAAGAAGCGCTATGCGGTCATAGATGAAGAGGGAAAAGTTATTACGCGCGGTCTGGAGATAGTCCGTAGGGACTGGTCGGAAATCGCGAAAGAAACACAAGCGCGAGTGTTGGAAACCATCTTAAAGCACGGGGACGTTGAGGAGGCTGTTAGGATTGTAAAAGAAGTCATCCAGAAGCTGGCAAATTATGAAATCCCTCCTGAGAAACTGGCCATCTACGAGCAAATTACTAGACCCTTGCATGAGTATAAGGCAATTGGTCCACATGTGGCGGTGGCTAAGAAACTTGCGGCGAAGGGTGTTAAAATCAAACCCGGCATGGTAATTGGATATATTGTGCTGAGGGGAGATGGTCCTATTTCCAATCGCGCCATATTAGCAGAGGAATATGATCCAAAGAAGCACAAATACGATGCTGAATACTATATCGAAAATCAGGTCTTACCCGCGGTATTGCGTATATTAGAGGGCTTTGGATACCGCAAAGAGGACTTGCGGTATCAGAAAACGAGACAGGTGGGATTGACGTCTTGGTTAAATATAAAGAAGTCGGGGACGGGCGGAGGCGGCGCAACCGTAAAATTTAAGTACAAAGGCGAGGAGAAAGAAGTTGACATCAGCAAAATCAAGAAAGTGTGGCGCGTTGGGAAAATGATCTCGTTCACTTATGATGAAGGTGGCGGCAAAACCGGTCGTGGCGCCGTATCGGAGAAAGATGCACCAAAGGAACTTTTACAGATGTTAGAGAAACAGAAGAAAGGAGGTGGCTCAGGCGGCGGATCGGAAAACCTGTATTTTCAGGGTGGAGGCGGGTCCATGGTATCAAGTGGAGAAGATATTTTCTCCGGCCTGGTGCCTATCCTGATTGAGTTAGAGGGCGACGTTAATGGCCACAGATTTTCGGTTCGTGGCGAAGGATATGGTGATGCTTCAAATGGCAAATTGGAGATCAAATTTATTTGCACCACGGGTCGCTTGCCGGTCCCGTGGCCGACCCTGGTAACAACGCTGAGTTATGGAGTCCAGTGCTTTGCGAAATACCCAGAGCACATGCGCCAGAATGACTTCTTTAAGAGCGCAATGCCGGATGGCTATGTTCAGGAGCGCACGATCAGCTTTAAAGAGGACGGAACCTACAAAACACGAGCAGAAGTTAAATTCGAAGGGGAGGCGCTTGTTAACCGGATCGACCTGAAGGGCTTAGAGTTCAAAGAGGATGGAAACATCTTGGGTCACAAACTTGAATACTCATTTAATAGTCACTATGTGTATATTACGGCGGATAAGAACAGGAATGGACTGGAGGCACAGTTTCGAATCCGGCATAATGTCGATGATGGTTCCGTCCAATTGGCGGACCATTACCAGCAAAATACCCCGATCGGGGAGGGCCCCGTGTTGTTACCTGAACAGCACTATTTAACCACCAATAGCGTTCTTTCTAAAGATCCCCAAGAACGCAGGGACCACATGGTTCTAGTTGAATTTGTGACTGCTGCCGGTTTGAGCTTAGGAATGGATGAATTGTACAAAAGCGGTGGAGGCAGCCACCATCACCACCACCATCATCACCATCACTCTTCCAAGAAATCAGGTTCATACTCTGGTTCCAAAGGCTCAAAACGGCGGATTCTGTAATAA"]},"metadata":{},"execution_count":9}]},{"cell_type":"markdown","metadata":{"id":"P9CNus6GmTNK"},"source":["Then you can just save it with the code below and inspect it as you wish!"]},{"cell_type":"code","metadata":{"id":"z4yo7Nc83Zex"},"source":["var fixedList []fasta.Fasta\n","fixedList = append(fixedList, fasta.Fasta{\"My fixed CDS sequence\", fixedSequence}) \n","fasta.Write(fixedList, \"fixedSequence.fasta\")"],"execution_count":null,"outputs":[]}]}